feat: KID and fnr generator/validator

tanettrimas opened the pull request:
Closes  #2

The following changes are applied in this commit:

* Refactored some of the methods used in mod11 out into helper lib, which can/is reused in mod10.js
* Added a mod10 lib
* Added a generator for kidnumber using mod10, either with a desired length, random or with a desired length and invoice/customer number e.g.

tanettrimas committed:
feat(kidnumber): first version of KID-nummer generator/validator using MOD10

The following changes are applied in this commit:

* Refactored some of the methods used in mod11 out into helper lib, which can/is reused in mod10.js
* Added a mod10 lib
* Added a generator for kidnumber using mod10, either with a desired length, random or with a desired length and invoice/customer number e.g.

tanettrimas commented:
@ergor Have a look if you so please 👍

tanettrimas committed:
chore: update readme

tanettrimas committed:
chore: fix indentation

ergor commented on src/app.js:
```diff
@@ -1,7 +1,23 @@
+	? generateKidMod10(kidNumberLength)
+	: generateKidMod10(kidNumberLength, kidNumberBase)
+	document.querySelector('.js-kid-nummer').innerText = kidNumber
+	kidNumberSpan.value = ''
+	kidNumberLengthSpan.value = ''
```
This comes down to preference, but I would like them to not be cleared. Say I want to generate a couple of kid numbers with length 15 but with different invoice numbers. I would have to retype 15 every time

tanettrimas replied:
That is true, I can fix that! :)

tanettrimas replied:
Fixed by b2218b7

ergor commented on src/mod10.js:
```diff
@@ -0,0 +1,44 @@
+ */
+export function mod10_generate(length, digits = gen(length)) {
+  const extraCiphers = gen(length - (digits.length + 1))
+  const mod11Base = [...digits, ...extraCiphers]
```
mod11 in mod10?

tanettrimas replied:
...ooops, typo! Will fix

tanettrimas replied:
Fixed by b2218b7

ergor commented on src/mod10.js:
```diff
@@ -0,0 +1,44 @@
+/**
+ * 
+ * @param {number} length number of digits excluding the control digit
+ * @param {number[]} weights (optional)
```
Mismatch between parameter and its doc

tanettrimas replied:
Yes, should have correct name and be specified that is used gen as default if parameter not included

tanettrimas replied:
Fixed by b2218b7

ergor commented on src/helpers/validateMod.js:
```diff
@@ -0,0 +1,24 @@
+ * @param {(digits: number[], weights: number[]) => number} modulo mod10 or mod11 function
+ * @param {'mod10' | 'mod11'} modType the type of modulo calculation used
+ */
+export default function validate(digits, weights, modulo, modType) {
```
Since we don't use this method directly but via `mod11_validate` and `mod10_validate`, we could do without the `modType` parameter.
Instead, we could require that weights are supplied, which would be the job of `mod11_validate` and `mod10_validate` to supply the correct ones.

tanettrimas replied:
Fixed by b2218b7

ergor reviewed (changes requested):
It is the right move to extract common functions between mod11 and mod10, but I think it would make sense to extract them into a singular `modCommon.js` or something, instead of having each function as a standalone file. All of the functions pertain to mod algorithms, and they should be grouped as such.

tanettrimas committed:
refactor: 💡 moved functions into modCommon++

From feedback on PR #17, functions in own files were moved into
modCommon.js. Typos are fixed and span in html are not removed from
generation because it can be a bit annoying.

tanettrimas committed:
docs: ✏️ Added missing jsdoc

tanettrimas committed:
fix: 🐛 Wrongfully removed input validation on account number

ergor commented on src/kid.js:
```diff
@@ -15,15 +16,29 @@ export function generateKidMod10(length, baseNum) {
+  return checkModulo(modulo, { mod10: mod10_generate, mod11: mod11_generate }, kidLength - 1)
+}
+
+function checkModulo(modulo, { mod10, mod11 }, ...opts) {
```
Isn't the `{ mod10, mod11 }` parameter redundant?
In the switch, you could call `mod10_generate` or `mod11_generate` directly

Also the name `checkModulo` suggests the return type is a boolean, I would rename it to applyModulo or something like that. The doc suggests it can return a boolean, but I don't see where in the function?

Finally, I think the doc for `modulo` is ambigous, as "the modulo" can be interpreted as meaning "the result of a modulo operation". It should be clear in the doc that it is for selecting which modulo function is applied.

tanettrimas replied:
It is not completely redundant as the mod10/mod11 is either mod10_generate or validate based on who is calling it. Therefore, checkModulo works in both scenarioes here. By calling generate directly in the switch, the validation would stop working and not be as dynamic. I agree with the nameChanges and doc-suggestions. Will fix that

tanettrimas replied:
checkModulo returns a boolean when mod10/11_validate is used.

tanettrimas replied:
Can confirm that all functions in the page works as expected. @ergor We should possibly add some unit tests before this project goes out of hand? :D

tanettrimas committed:
Merge branch 'master' of https://github.com/tanettrimas/YesCommons into feat/kidnumber

ergor commented on src/kid.js:
```diff
@@ -39,6 +55,12 @@ function checkModulo(modulo, { mod10, mod11 }, ...opts) {
+/**
+ * 
+ * @param {String} kid The kid number
+ * @param {String} modulo The modulo 
```
`{String} modulo The modulo` is ambigous. See other comment

tanettrimas replied:
fixed by f9aa173

ergor commented on src/kid.js:
```diff
@@ -1,6 +1,14 @@
+ * @param {Object} params
+ * @param {number} params.length Length of the kidnumber
+ * @param {String[]} params.baseNum The basenumbers
+ * @param {String} params.modulo The modulo
```
`{String} modulo The modulo` is ambigous. See other comment

tanettrimas replied:
fixed by f9aa173

ergor reviewed (changes requested):
Good refactor :+1:  Have you tested the old tools relying on mod11 for regressions? 
Also, see comments.

tanettrimas committed:
fix: 🐛 mod10_generate was returning invalid kid

There was a bug introduced by a typo in modCommon.js#generate, was fixed
there by simplifying the method a bit when calling the provided modulo.
Now, it only calls the modulo function with the modbase instead of
calculating the weights beforehand.

tanettrimas committed:
docs: ✏️ Clearify some of the jsdocumentation and renaming

tanettrimas committed:
feat: 🎸 generator for norwegian socialsec number

refactor: 💡 made mod10 and mod11 generate more adaptable

By adding weights as an optional parameter, it can be used also when
generating social security numbers as they are using a fixed set of
weights.

docs: ✏️ Added missing jsdoc to the mod generate

feat: generator for norwegian socialsec number

* Added new for generation of social security numbers. Either a normal fødselsnummer,
d-nummer, h-nummer and fh-nummer. Ref: https://no.wikipedia.org/wiki/F%C3%B8dselsnummer

* Added some date-helper functions which is primarily used for social security number validation.

refactor: 💡 Re-structurized the files for better maintenance

refactor: 💡 small renaming variable changes and cleanup

Made sure that variable names are a bit clearer and that validation of
Dnumber for example are only done once and not duplicated all over

feat: 🎸 added validation and couple of helpers

Now you can validate dnumber, fnumber, hnumber and fhnumber. It also is
printed pretty with a couple of helpers.

feat: 🎸 finish fnr-validator/generator

tanettrimas committed:
Merge pull request #1 from tanettrimas/feat/fnr

feat: 🎸 generator for norwegian socialsec number

tanettrimas commented:
@ergor I merged the fnr feature branch now :)

tanettrimas committed:
Merge branch 'master' into feat/kidnumber
